book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -0,0 +1,442 @@
+
+If we perform all *writes* synchronously and allow reads to be either
+synchronous or asynchronous, we ensure that writes cannot race each other.
+Asynchronous reads are guaranteed to read at least the state present at the
+time the request was processed, or a later state.
+
+In summary:
+
+- **Sled reads** may be done synchronously (in `call`) or asynchronously (in
+  the `Future`), depending on the context;
+
+- **Sled writes** must be done synchronously (in `call`).
+
+## In-memory data structures
+[in-memory]: #in-memory
+
+At a high level, the in-memory data structures store a collection of chains,
+each rooted at the highest finalized block. Each chain consists of a map from
+heights to blocks. Chains are stored using an ordered map from difficulty to
+chains, so that the map ordering is the ordering of best to worst chains.
+
+- XXX fill in details on exact types
+
+- **Chain**: `(im::OrdMap<block::Height, Arc<Block>>, HashSet<Nullifier>, HashSet<block::Hash>, HashSet<Anchor>, HashSet<UTXO>)`
+  - Ord impl is ordered by work
+  - push => add a block to the end of a chain, does contextual verification
+  checks, extracts info from block for extra data sets
+  - pop => remove the lowest block, remove references to contents from block in various extra data sets (nullifiers, hashes, etc)
+  - fork => create a new chain fork based on a given block(hash) within
+  another chain, calls push repeatedly
+- **ChainSet**: `(BTreeSet<Chain>, BTreeMap<block::Height, Arc<Block>>)`
+  - `fn finalize(&mut self) -> Arc<Block>`
+  - `fn commit_block(&mut self, block: Arc<Block>) -> Result<(), Err>`
+    - iterate through chains, if the parent of `block` is the tip of any
+    chains, try to push block onto that chain
+    - if not, iterate through chains, checking if the parent is contained in
+    that chain
+    - if so, fork the chain at the parent and try to push `block` onto that
+    fork and add newly extended chain if the push succeeds
+    - if not, queue block, prune queued blocks that are below the reorg limit
+
+
+- XXX work out whether we should store extra data (e.g., a HashSet of UTXOs
+  spent by some block etc.) to speed up checks.
+
+When a new block extends the best chain past 100 blocks, the old root is
+removed from the non-finalized state and committed to the finalized state.
+
+## Committing non-finalized blocks
+
+If the parent block is not committed, add the block to an internal queue for
+future processing.
+
+Otherwise, attempt to perform contextual validation checks and the commit the
+given block to the state. The exact list of contextual validation checks will
+be specified in a later RFC. If contextual validation checks succeed, commit
+the new blocks to the non-finalized state as described below. Next, if the
+resulting non-finalized chain is longer than 100 blocks, the oldest block is
+now past the reorg limit. Remove it from the non-finalized state and commit
+it as a finalized block, as described in the next section.
```

Comment:
Let's make sure we handle pruning side-chains correctly here, and in `finalize` above.

We want to commit the root block from the best chain. If a chain doesn't contain that block, we want to discard *the whole chain*, because it can't possibly be a future best tip.